Tool to generate a HDF table of sampling interval coefficients

SeiyaNozaki opened the pull request:
This PR is a continuation of #255.
I added a new tool and class to create sampling interval coefficients tables, which is used for `TimeSamplingCorrection`.

This tool has three stages:
- count pulse peak on each capacitor
  -   we usually use 1 million events, so better to submit a subrun-wise job to save time. It seems some modules have some problems with this method (almost synchronization between test pulse and the first capacitor id), but this trouble appears randomly for all of the modules. So we take 2-3 runs for a single gain to generate sampling interval coefficients. 

- stack single fits file, convert into sampling interval coefficients, calculate charge resolution with new coefficients for the verification
  - stack multiple fits files generated by subrun-wise jobs and convert the stucked peak counts into sampling interval coefficients for each run. Then,  charge resolutions are calculated with new coefficients for the verification. If a given module has trouble in a specific run, charge resolution would be not good than expected. So the final coefficient values would be what achieved the best resolutions. The final coefficients and used run for the generation of coefficients on each module are saved for each gain.

-  merge coefficients of both gains into a single HDF table
   - merge output files of the previous stage for each gain. The generated HDF file should be compatible with `TimeSamplingCorrection` already implemented in lstchain.
  https://github.com/cta-observatory/cta-lstchain/blob/master/lstchain/calib/camera/time_sampling_correction.py#L14

SeiyaNozaki committed:
add a tool to create sampling interval coefficients

SeiyaNozaki committed:
Merge remote-tracking branch 'upstream/master' into samp_interval_coefficient

maxnoe commented on lstchain/calib/camera/sampling_interval_coefficient_calculate.py:
```diff
@@ -0,0 +1,170 @@
+        The SamplingIntervalCalculate class to create a sampling interval coefficient table for LST readout system using chip DRS4.
+    """
+    def __init__(self):
+        self.peak_count = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL])
```
`zeros` creates a float array by default, counts should be integers, right? So specify `dtype=np.int64` or another appropriate int dtype.

SeiyaNozaki replied:
thanks, I set the proper data type (uint16 is enough).

maxnoe commented on lstchain/calib/camera/sampling_interval_coefficient_calculate.py:
```diff
@@ -0,0 +1,170 @@
+    """
+    def __init__(self):
+        self.peak_count = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL])
+        self.fc_count = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL])
```
As above

maxnoe commented on lstchain/calib/camera/sampling_interval_coefficient_calculate.py:
```diff
@@ -0,0 +1,170 @@
+            self.fc_count[np.arange(N_PIXELS), fc % N_CAPACITORS_CHANNEL] += 1
+            
+
+    def stuck_single_sampling_interval(self, file_list, gain):
```
What does `stuck` mean here? Are you sure that's the right word?

SeiyaNozaki replied:
sorry, it is a typo... it should be 'stack'

maxnoe commented on lstchain/calib/camera/sampling_interval_coefficient_calculate.py:
```diff
@@ -0,0 +1,170 @@
+            
+            if gain_in_file == gain:
+                if not run_id in self.peak_count_stuck.keys():
+                    self.peak_count_stuck[run_id] = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL])
```
dtype

maxnoe commented on lstchain/calib/camera/sampling_interval_coefficient_calculate.py:
```diff
@@ -0,0 +1,170 @@
+            if gain_in_file == gain:
+                if not run_id in self.peak_count_stuck.keys():
+                    self.peak_count_stuck[run_id] = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL])
+                    self.fc_count_stuck[run_id] = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL])
```
dtype?

maxnoe commented on lstchain/calib/camera/sampling_interval_coefficient_calculate.py:
```diff
@@ -0,0 +1,170 @@
+            self.sampling_interval_coefficient[run_id] = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL + N_SAMPLES])
+
+            for pixel in range(N_PIXELS):
+                self.sampling_interval_coefficient[run_id][pixel, :N_CAPACITORS_CHANNEL] = \
```
Please use implicit line continuations inside parentheses instead of these line breaks with `\`.

Also, it looks like this loop is not necessary but could be avoided by using the `axis` keyword for the numpy functions.

SeiyaNozaki replied:
thanks for the comment. done.
(Concerning the loop, please let me know if there is a more straightforward way...!)

maxnoe commented on lstchain/calib/camera/sampling_interval_coefficient_calculate.py:
```diff
@@ -0,0 +1,170 @@
+        r1_sample_end = r0_r1_calibrator.r1_sample_end.tel[tel_id]
+
+        # Check pulse
+        pulse_event_flag = np.sum(np.max(waveform[gain], axis=1) > 100) > 1800
```
This needs more context and probably the unnamed constants here should be transformed into either global constants or configurable traitlets.

SeiyaNozaki replied:
added contexts, and set global constants for them

maxnoe commented on lstchain/calib/camera/sampling_interval_coefficient_calculate.py:
```diff
@@ -0,0 +1,170 @@
+
+                
+    def set_charge_array(self, gain):
+        self.charge_array_before_corr = np.zeros([N_PIXELS, 60000])
```
Why 60000?

SeiyaNozaki replied:
I modified related codes. If `max_events` is given, this given value will be used to generate a new array.
If not (`max_events` is `None` by default), just use a large enough number (55000) to cover all events in a single subrun.
(A single subrun usually contains ~53000 events.)
I'm sure it is not the best way... so I'm very happy if you have any good idea...

maxnoe replied:
You can actually get the number of events by doing `len(source.multi_file)`, I will also add `__len__` to the `LSTEventSource`, but that will require a new release.

A more flexible solution would be to use a python list and append to it, converting to a 2d numpy array at the end.

maxnoe commented on lstchain/calib/camera/sampling_interval_coefficient_calculate.py:
```diff
@@ -0,0 +1,170 @@
+        self.charge_reso_array_before_corr = np.zeros(N_PIXELS)
+
+        for run_id in self.peak_count_stuck.keys():
+            self.charge_array_after_corr[run_id] = np.zeros([N_PIXELS, 60000])
```
same as above

maxnoe commented on lstchain/calib/camera/sampling_interval_coefficient_calculate.py:
```diff
@@ -0,0 +1,170 @@
+
+__all__ = ['SamplingIntervalCalculate']
+
+N_PIXELS = 1855
```
you can import these from `ctapipe_io_lst.constants.`

SeiyaNozaki replied:
done!

maxnoe commented on lstchain/tools/lstchain_create_sampling_interval_coefficient_file.py:
```diff
@@ -0,0 +1,248 @@
+        self.log.debug('start peak counting')
+
+        for i, event in enumerate(self.eventsource):
+            if event.index.event_id % 500 == 0:
```
Use `tqdm` for a proper progress bar instead.

SeiyaNozaki replied:
ok, I didn't understand `tqdm` before:) Now `tqdm` is used during even loop.

maxnoe replied:
Then this line is not needed anymore

SeiyaNozaki committed:
fix typo

SeiyaNozaki committed:
set uint16 as dtype for some parameters

SeiyaNozaki committed:
import ctapipe_io_lst.constants for global parameters

SeiyaNozaki committed:
add general constants and explanations

SeiyaNozaki committed:
remove line breaks with backslash

SeiyaNozaki committed:
use tqdm

SeiyaNozaki committed:
import tqdm

maxnoe commented on lstchain/tools/lstchain_create_sampling_interval_coefficient_file.py:
```diff
@@ -0,0 +1,262 @@
+from lstchain.calib.camera.sampling_interval_coefficient_calculate import SamplingIntervalCalculate
+from lstchain.paths import run_info_from_filename
+
+class SamplingIntervalCoefficientHDFWriter(Tool):
```
This seems just to be three tools in one that don't really share code and even use different traitlets. I think really doing three smaller tools would be better and less confusing (since the help will only show the options for each step, not for all steps).

SeiyaNozaki replied:
I see. Actually, I just wanted to gather all of the procedures in a single file. Is it possible to put three smaller tools in a single file and call one of the tools when executing the command...? or one file per one tool?

maxnoe replied:
You can do multiple Tools per file and have different entrypoints, but I would just make three files

FrancaCassol commented on lstchain/tools/lstchain_create_sampling_interval_coefficient_file.py:
```diff
@@ -0,0 +1,262 @@
+    lstchain_create_sampling_interval_coefficient_file --help
+    """
+
+    name = "SamplingIntervalCoefficientHDFWriter"
```
Hi Seiya,
if I understand correctly you are writing fits files, instead of a HDF5 file.  
Would be it possible to write indeed a h5 file? (we decided to prefer this format)

FrancaCassol replied:
Forget my comment, I see that the final file is in h5 and also that you put together the gains at the ends. 
Perhaps a bit misleading the fact that you have single tool for the three steps

FrancaCassol commented on lstchain/tools/lstchain_create_sampling_interval_coefficient_file.py:
```diff
@@ -0,0 +1,262 @@
+        default_value="*.fits"
+    ).tag(config=True)
+
+    gain = traits.Int(
```
Is there a reason to sparate HG and LG? 
It would be better to have all the data in a single file, so better to process them together if possible

SeiyaNozaki committed:
refactoring (separate procedures into two Tools and files, add a function to plot results)

SeiyaNozaki commented:
Thank you for the comments, @maxnoe and @FrancaCassol !
I refactored the codes to make them simple and easy to understand.
Now there are two steps (`lstchain_create_peak_count_file_for_sampling_interval_calib` and `lstchain_create_sampling_interval_coefficient_file`): the first one is to count the pulse peak on each capacitor and the second is to stack, convert and verify the sampling interval coefficients.
The summary of the procedure is below:
![image](https://user-images.githubusercontent.com/31307823/116776297-8f44b280-aaa2-11eb-83dc-2bcc57bbef62.png)

I also added a function to produce a result pdf file like below:
[sampling_interval_coefficient_results_HG.pdf](https://github.com/cta-observatory/cta-lstchain/files/6409153/sampling_interval_coefficient_results_HG.pdf)

It would be very appreciated to have a look again.

SeiyaNozaki commented:
Hi @FrancaCassol @maxnoe,
Sorry for the delay, but came back again.
I'm happy if you can check the updated codes again when you have time!

FrancaCassol commented on lstchain/calib/camera/sampling_interval_coefficient_calculate.py:
```diff
@@ -0,0 +1,283 @@
+    def plot_results(self, verify_data_path, output_file):
+
+        plt.rcParams['font.size']=15
+        camera = CameraGeometry.from_name('LSTCam-002')
```
This is an outdated camera (we have now version 4).
Better to use:
```
from ctapipe_io_lst import load_camera_geometry
camera = load_camera_geometry()
camera = camera.transform_to(EngineeringCameraFrame())
```

FrancaCassol commented on lstchain/tools/lstchain_create_sampling_interval_coefficient_file.py:
```diff
@@ -20,21 +25,11 @@ class SamplingIntervalCoefficientHDFWriter(Tool):
-    input_fits = traits.Path(
-        help="Path to the generated peak count fits files",
+    input_peak_count_fits = traits.Path(
+        help="Path to the generated peak count fits files by lstchain_create_peak_count_file_for_sampling_interval_calib",
```
Hi @SeiyaNozaki ,

I do not see this tool in your PR, am I wrong?

maxnoe commented on lstchain/tools/lstchain_create_sampling_interval_coefficient_file.py:
```diff
@@ -0,0 +1,133 @@
+        self.sampling_interval_coefficient_calculate.convert_to_samp_interval_coefficient()
+        self.sampling_interval_coefficient_calculate.set_charge_array()
+
+        for i, event in enumerate(tqdm(
```
No need for enumarete, the event already has `event.count`
